wfe/wfe.go

```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
 	}
 }
 
+func (wfe *WebFrontEndImpl) KeyRollover(
+	ctx context.Context,
+	logEvent *requestEvent,
+	response http.ResponseWriter,
+	request *http.Request) {
+	// Extract and parse outer JWS, and retrieve account
+	body, outerHeaderURL, key, prob := wfe.verifyPOST(ctx, logEvent, request, wfe.lookupJWK)
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	existingAcct, prob := wfe.getAcctByKey(key)
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	// Extract inner JWS
+	parsedInnerJWS, err := wfe.parseJWS(string(body))
+	if err != nil {
+		wfe.sendError(acme.MalformedProblem(err.Error()), response)
+		return
+	}
+
+	innerHeaderURL, ok := parsedInnerJWS.Signatures[0].Header.ExtraHeaders[jose.HeaderKey("url")].(string)
+	if !ok || len(innerHeaderURL) == 0 {
+		wfe.sendError(acme.MalformedProblem("JWS header parameter 'url' required for inner JWS."), response)
+	}
+	if innerHeaderURL != outerHeaderURL {
+		wfe.sendError(acme.MalformedProblem("JWS header parameter 'url' differ for inner and outer JWS."), response)
+	}
+
+	newPubKey, prob := wfe.extractJWK(request, parsedInnerJWS)
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	// Copied from verifyJWS:
+	// TODO(@cpu): `checkAlgorithm()`
+
+	innerPayload, err := parsedInnerJWS.Verify(newPubKey)
+	if err != nil {
+		wfe.sendError(acme.MalformedProblem("JWS verification error"), response)
+		return
+	}
+
+	var innerContent struct {
+		Account string `json:"account"`
+		OldKey  jose.JSONWebKey `json:"oldKey"`
+	}
+	err = json.Unmarshal([]byte(innerPayload), &innerContent)
+	if err != nil {
+		wfe.sendError(acme.MalformedProblem("Error unmarshaling key roll-over innermost JSON body"), response)
+		return
+	}
+
+	// Check account ID
+	prefix := wfe.relativeEndpoint(request, acctPath)
+	if !strings.HasPrefix(innerContent.Account, prefix) {
+		wfe.sendError(acme.MalformedProblem("Key ID (account) in innermost JSON body missing expected URL prefix"), response)
+		return
+	}
+	accountID := strings.TrimPrefix(innerContent.Account, prefix)
+	if accountID == "" {
+		wfe.sendError(acme.MalformedProblem("No key ID (account) in innermost JSON body"), response)
+		return
+	}
+	if accountID != existingAcct.ID {
+		wfe.sendError(acme.MalformedProblem("Key roll-over innermost JSON contains wrong account ID"), response)
+		return
+	}
+
+	// Verify inner key
+	if !keyDigestEquals(innerContent.OldKey, *existingAcct.Key) {
+		wfe.sendError(acme.MalformedProblem("Key roll-over innermost JSON contains wrong old key"), response)
+		return
+	}
+
+	// Ok, now change account key
+	err, otherAccount := wfe.db.ChangeAccountKey(existingAcct, newPubKey)
+	if err != nil {
```

Comment:
This is where you will want to check whether `err` is of the specific type that you defined for a key conflict, and handle that accordingly.